add remote check to secret and variable commands

wingleung opened the pull request:
Fixes #4688

wingleung committed:
WIP add remote check to secret and variable commands

wingleung committed:
add tests

wingleung committed:
Merge remote-tracking branch 'upstream/trunk' into add-remote-check-to-secret-and-variable

Jorslope commented on pkg/cmd/secret/delete/delete.go:
```diff
@@ -103,6 +110,11 @@ func removeRun(opts *DeleteOptions) error {
 		}
+
+		err = cmdutil.ValidateHasOnlyOneRemote(opts.HasRepoOverride, opts.Remotes)
+		if err != nil {
```
Si

andyfeller commented:
@wingleung : thank you yet again for your time and effort to improve the GitHub CLI and your patience with my slow follow up! 🙇 

reviewing this right now, wanting to offer feedback today

andyfeller commented on pkg/cmdutil/errors.go:
```diff
@@ -57,6 +58,21 @@ func MutuallyExclusive(message string, conditions ...bool) error {
 	return nil
 }
 
+func ValidateHasOnlyOneRemote(hasRepoOverride bool, remotes func() (ghContext.Remotes, error)) error {
+	if !hasRepoOverride && remotes != nil {
+		remotes, err := remotes()
+		if err != nil {
+			return err
+		}
+
+		if remotes.Len() > 1 {
+			return fmt.Errorf("multiple remotes detected %v. please specify which repo to use by providing the -R or --repo argument", remotes)
+		}
+	}
+
+	return nil
+}
```
@wingleung : reflecting on https://github.com/cli/cli/issues/4688#issuecomment-1016708631 versus the code here, I think the original approach of prompting the user regarding which remote to use feels like the right choice still, erring if they don't pick a remote or if they requested [no prompting](https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/iostreams/iostreams.go#L250-L256).

This would mean we don't need users to provide `-R,--repo` override as the previous workaround involved.

### Other commands that handle remotes

https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/cmd/codespace/create.go#L143-L155

https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/cmd/extension/manager.go#L691-L716

https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/cmd/pr/merge/merge.go#L403-L431

williammartin replied:
Gentle nudge on this @wingleung, don't want it to get stale.

@andyfeller if I read this correctly it's a breaking change for the following **non-interactive** script:

```
gh repo create org/repo
gh repo fork org/repo
gh repo clone me/fork <-- sets the default to org/repo
gh secret set FOO BAR
```

Previously this would have put the secret into upstream (the bug), but now it will error. Just want to make sure we are clear and can communicate this.

andyfeller replied:
> @andyfeller if I read this correctly it's a breaking change for the following non-interactive script

@williammartin : yeah; if we need to fallback to prompting when in interactive mode and encounter 2+ remotes, then this will break in non-interactive scripts.

**Is there a different way you'd suggest for addressing this than than https://github.com/cli/cli/issues/4688#issuecomment-1016708631?**

The security concern on this one resonates with me strongly.  If we're accidentally causing users to set secrets and variables on the wrong repositories, then I think we should fail when it is ambiguous in non-interactive mode. 🤷

williammartin replied:
I can't see any way to avoid it being a breaking change while also helping our users avoid a footgun. I'm a bit concerned about how many people it might break though.

https://github.com/search?type=code&q=%22gh+secret+set%22

More than 1.7 public results, and any of these that are run non-interactively against forks will be broken.

wingleung replied:
I've been very busy at work so apologies for the late reply.

OK, so if I understand correctly, there are 3 flows we want to cover, 2 of which are already covered in this PR

### ✅ flow 1: command with 1 remote
just use that single remote and proceed with the code

### ✅ flow 2: command with multiple remotes in non-interactive mode
error out but provide documentation on how to force a remote with `--repo`.

> This would mean we don't need users to provide -R,--repo override as the previous workaround involved.

in this case we would still need to support `-R` and `--repo` overrides, otherwise non-interactive flow will get stuck, or am I not interpreting it correctly? 😅 

```bash
multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
```

**use cases**

- `gh secret set MYSECRET --body "$ENV_VALUE”`
- `gh variable set MYVARIABLE --body "$ENV_VALUE”`

### 🚫 flow 3: command with multiple remotes in interactive mode
show a prompt to select 1 remote repo from multiple available remote repos using a search prompt 

**use cases**

- `gh secret list`
- `gh secret set MYSECRET`
- `gh variable set MYVARIABLE`

Is that correct?

wingleung replied:
### Breaking changes

we could do a minor release to introduce a warning for the v2 cli and mention this would be a breaking change in v3? Meanwhile working on a v3 PR

wingleung replied:
nice to see how remotes are being handle in other places, maybe in another ticket it could be looked into on how to merge all those strategies into 1 helper function.

- default to `origin`, fallback to first remote 👉 https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/cmd/extension/manager.go#L691-L716

- default remote set to configured `set-default` remote value with a fallback to the first remote and prompt for user input on which remote to pick 👉 https://github.com/cli/cli/blob/e36d8f103d5e1fb3989d9d5aefdae36b01fb88d4/pkg/cmd/codespace/create.go#L143-L174

just to prevent this PR from getting too big or out of scope, I'd suggest to do this in another ticket 😄

andyfeller replied:
> I've been very busy at work so apologies for the late reply.

@wingleung : I just want to say you owe us no apology; thank you for yet again working to improve `gh` ❤️

andyfeller replied:
I think this summarizes the changes adequately, yes:

> ✅ flow 1: command with 1 remote
> ✅ flow 2: command with multiple remotes in non-interactive mode
> 🚫 flow 3: command with multiple remotes in interactive mode

Along with the actual changes to the command, we will want to make sure to call this out in the release notes to explain the motivation and hopefully how this is a one-time issue for non-interactive situations.

wingleung replied:
https://github.com/cli/cli/pull/9083/commits/81e7ecdfc135854c2609f769c4b47251da80b78e added a prompt for interactive mode on following commands
- `gh secret set MYSECRET`
- `gh variable set MYVARIABLE`

the other `get`, `delete`, `list` commands don't seem to support interactive mode.

TODO: add tests

wingleung committed:
Merge remote-tracking branch 'upstream/trunk' into add-remote-check-to-secret-and-variable

wingleung committed:
prompt to select a repo in interactive mode

wingleung committed:
fix lint

andyfeller commented on pkg/cmdutil/repo_override.go:
```diff
@@ -68,3 +70,34 @@ func OverrideBaseRepoFunc(f *Factory, override string) func() (ghrepo.Interface,
+	var defaultRepo string
+	var remoteArray []string
+
+	if remotes, _ := remotes(); remotes != nil {
+		if defaultRemote, _ := remotes.ResolvedRemote(); defaultRemote != nil {
+			// this is a remote explicitly chosen via `repo set-default`
+			defaultRepo = ghrepo.FullName(defaultRemote)
+		} else if len(remotes) > 0 {
+			// as a fallback, just pick the first remote
+			defaultRepo = ghrepo.FullName(remotes[0])
+		}
+
+		for _, remote := range remotes {
+			remoteArray = append(remoteArray, ghrepo.FullName(remote))
+		}
+	}
+
+	baseRepoInput, errInput := survey.Select("Select a base repo", defaultRepo, remoteArray)
+	if errInput != nil {
+		return baseRepo, errInput
+	}
```
Everything from this logic tracks for me.  My main question is whether situations where `gh repo set-default` has explicitly set a remote / base should just use it rather than falling back to the prompting logic 🤔

wingleung replied:
I was in doubt here, it would kind of feel like forced magic here. Yes, you might set the default but there could be another repo you want to create secrets on, then you need to remember to override with the `--repo` flag because there would be no way in interactive mode to switch repos once you have a `set-default` one. I thought being explicit and transparent to the user would be the safer bet here

andyfeller replied:
First, I want to say thank you for taking security so seriously 🤗 at GitHub!  Our Engineering leader `@mph4` constantly says _"Raise your ✋ if you're on the Security team"_ and I always do.

> Yes, you might set the default but there could be another repo you want to create secrets on, then you need to remember to override with the --repo flag because there would be no way in interactive mode to switch repos once you have a set-default one.

[`gh repo set-default`](https://cli.github.com/manual/gh_repo_set-default) already states that it is used to determine the base repo for creating repository and environment secrets:

> This command sets the default remote repository to use when querying the GitHub API for the locally cloned repository.
> 
> gh uses the default repository for things like:
> 
> - viewing and creating pull requests
> - viewing and creating issues
> - viewing and creating releases
> - working with GitHub Actions
> - adding repository and environment secrets

I'm sympathetic to this concern, however I believe an expectation has been set that prompting when there is a base repo set is counter to this and is at the edge of shared responsibility.

wingleung replied:
oooh I see, thanks for explaining it to me, I should've seen it in the docs. Even if it would be counter intuitive for some, that would be for another discussion. Given the guidelines of `set-default` I completely agree we can just take the default remote and use that.
PR updated, I'll look into add tests soon!

andyfeller replied:
Again, thank you for being intentional and being open to discussion! 🙇

wingleung committed:
use default set remote if available

wingleung committed:
Merge remote-tracking branch 'upstream/trunk' into add-remote-check-to-secret-and-variable

wingleung committed:
add todos for tests

andyfeller commented on pkg/cmd/variable/set/set.go:
```diff
@@ -161,6 +175,26 @@ func setRun(opts *SetOptions) error {
 		if err != nil {
 			return err
 		}
+
+		err = cmdutil.ValidateHasOnlyOneRemote(opts.HasRepoOverride, opts.Remotes)
+		if err != nil {
+			defaultRepo := cmdutil.DefaultRepo(opts.Remotes)
+
+			if defaultRepo != nil {
+				baseRepo = defaultRepo
+			} else if opts.Interactive {
+				selectedRepo, errSelectedRepo := cmdutil.PromptForRepo(baseRepo, opts.Remotes, opts.Prompter)
+
+				if errSelectedRepo != nil {
+					return errSelectedRepo
+				}
+
+				baseRepo = selectedRepo
+			} else {
+				return err
+			}
+		}
+
```
**Taking a step back from the code, is the following the logical order that repository should be identified?**

1. If a user specifies `--owner` this is an organization secret; otherwise it is a repository secret of some sort

1. If a user specifies `--repo` this has the highest precedence because the user is explicitly telling us what to use

1. Otherwise, we figure out the repo based on the remotes, dealing with various scenarios
   1. If there are multiple remotes but we cannot interact, error
   2. If there are multiple remotes and we can interact, prompt the user and maybe mention they can avoid this in the future with either using `--repo ...`
   3. If there is 1 remote, we use it

andyfeller replied:
### Discussing with @williammartin ✨ 

@wingleung : firstly, allow me to apologize for confusion on my part; Will has highlighted the fact that `gh.resolved` flag on the remote that provides the base repo is actually contributing to the challenge we're trying to undo here.

During our conversation, Will highlighted several things including:

1. Remove the line about `gh repo set-default` as this is technically the problem we're running into
2. This PR should explicitly call out in `gh repo set-default` that it IS NOT used for repository + repository environment
3. In the remote prompt ordering, the upstream should be first as we know upset users will want to hit enter
4. With these changes, we need to ensure it is called out in release notes and mention the community should follow up in this issue about interest in a configuration option
5. Will also called out a larger problem regarding the inconsistent handling of repository defaults that might necessitate further changes

wingleung replied:
> 1. If a user specifies `--owner` this is an organization secret; otherwise it is a repository secret of some sort

I think you meant `--org` instead of `--owner`, if so ✅ 

> 2. If a user specifies `--repo` this has the highest precedence because the user is explicitly telling us what to use

I think number 2 would be a secret or variable set with the `--env` flag, to set variables and secrets on a deployment environment level. which would be place between `--org` and `--repo` in terms of scope level

wingleung replied:
> During our conversation, Will highlighted several things including:
> 
> 1. Remove the line about `gh repo set-default` as this is technically the problem we're running into

OK, reverted the last commit, it will still select the default in the select prompt if there is a default but submitting is up to the user

> 2. This PR should explicitly call out in `gh repo set-default` that it IS NOT used for repository + repository environment

Idd, it would do the exact opposite to what's in the docs, so the guidelines for `gh repo set-default` should change as well, added a proposal `### NOTE: gh does not use the default repository for managing repository and environment secrets or variables.`. Let me know if you see any other places where we can add this note

> 3. In the remote prompt ordering, the upstream should be first as we know upset users will want to hit enter

We piggyback on `Remotes` which is doing sorting by itself using the `Less` function and the `sort` library from golang 👉 https://github.com/cli/cli/blob/95a2f95f75f4b143699d87294788210ffb558248/context/remote.go#L60-L78 so the order is always `upstream` first even though the default selected option might be the `origin` one. Or do you mean we should always select the first option which is the `upstream` option? That would be not be as intuitive because I would expect the `origin` to be the default option 

> 4. With these changes, we need to ensure it is called out in release notes and mention the community should follow up in this issue about interest in a configuration option

👍 

> 5. Will also called out a larger problem regarding the inconsistent handling of repository defaults that might necessitate further changes

I agree, I also mentioned something similar here 👉  https://github.com/cli/cli/pull/9083#discussion_r1701547993. problem is, it's not just a feature or a change to a command, it's refactoring with a higher level overview in mind. Is that something I can kickstart in https://github.com/cli/cli/discussions/categories/ideas? Or do you prefer to have this discussion internally first?

wingleung committed:
Revert "use default set remote if available"

This reverts commit 5798eb7c4b40561afff241a61ae59285e025d811.

wingleung committed:
update docs for set-default

NOTE: gh does not use the default repository for managing repository and environment secrets or variables.

wingleung committed:
Merge remote-tracking branch 'upstream/trunk' into add-remote-check-to-secret-and-variable

wingleung committed:
update setdefault test

andyfeller commented:
## Manual testing with forked repo

The following scenarios are exercising changes from the PR on a fork of a repo, which should require explicitly setting the repo because of multiple remotes.

1. Building local `gh` for testing:

   ```shell
   andyfeller@Andys-MBP:cli/cli-trunk ‹add-remote-check-to-secret-and-variable›$ make
   go build -trimpath -ldflags "-X github.com/cli/cli/v2/internal/build.Date=2024-08-23 -X github.com/cli/cli/v2/internal/build.Version=v2.55.0-13-g97b10370 " -o bin/gh ./cmd/gh
   ```
   
2. Creating upstream repo for testing
   
   
   ```shell
   andyfeller@Andys-MBP:cli/cli-trunk ‹add-remote-check-to-secret-and-variable›$ ./bin/gh repo create --add-readme --description "Upstream repo used in testing GitHub CLI secret and variable behaviors" --disable-issues --disable-wiki --public andyfeller/gh-testing-01 
   ✓ Created repository andyfeller/gh-testing-01 on GitHub
     https://github.com/andyfeller/gh-testing-01
   ```

3. Creating downstream fork for testing
   
   ```shell
   andyfeller@Andys-MBP:cli/cli-trunk ‹add-remote-check-to-secret-and-variable›$ cd ~/Documents/workspace/tinyfists 
   andyfeller@Andys-MBP:workspace/tinyfists $ ~/Documents/workspace/cli/cli-trunk/bin/gh repo fork andyfeller/gh-testing-01 --org tinyfists --clone 
   ✓ Created fork tinyfists/gh-testing-01
   Cloning into 'gh-testing-01'...
   remote: Repository not found.
   fatal: repository 'https://github.com/tinyfists/gh-testing-01.git/' not found
   Cloning into 'gh-testing-01'...
   remote: Enumerating objects: 3, done.
   remote: Counting objects: 100% (3/3), done.
   remote: Compressing objects: 100% (2/2), done.
   remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
   Receiving objects: 100% (3/3), done.
   From https://github.com/andyfeller/gh-testing-01
    * [new branch]      main       -> upstream/main
   ✓ Cloned fork
   ! Repository andyfeller/gh-testing-01 set as the default repository. To learn more about the default repository, run: gh repo set-default --help
   andyfeller@Andys-MBP:workspace/tinyfists $ cd gh-testing-01 
   ```

### `gh secret set`

4. Confirming repo with multiple remotes forces user to choose
   
   ```shell
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret set TEST001 --body "wut"
   multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
   ```

5. Confirming setting secret for upstream works

   ```shell
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret set TEST001 --body "wut" --repo tinyfists/gh-testing-01
   ✓ Set Actions secret TEST001 for tinyfists/gh-testing-01
   ```
   
6. Confirming setting secret for downstream works

   ```shell
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret set TEST002 --body "wut" --repo andyfeller/gh-testing-01
   ✓ Set Actions secret TEST002 for andyfeller/gh-testing-01
   ```

### `gh secret list`

7. Confirming repo with multiple remotes forces user to choose
   
   ```shell
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret list                                                    
   multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
   ```

8. Confirming listing secret for upstream works
   
   ```shell
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret list --repo tinyfists/gh-testing-01
   NAME     UPDATED               
   TEST001  less than a minute ago
   ```
   
9. Confirming listing secret for upstream works

   ```shell
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret list --repo andyfeller/gh-testing-01
   NAME     UPDATED               
   TEST002  less than a minute ago
   ```

### `gh secret delete`

10. Confirming repo with multiple remotes forces user to choose
   
    ```shell
    andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret delete TEST001
    multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
    ```

11. Confirming deleting non-existent secret for upstream works as expected
    
    ```shell
    andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret delete TEST001 --repo andyfeller/gh-testing-01
    failed to delete secret TEST001: HTTP 404 (https://api.github.com/repos/andyfeller/gh-testing-01/actions/secrets/TEST001)
    ```
    
12. Confirming deleting secret for downstream works as expected

    ```shell
    andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh secret delete TEST001 --repo tinyfists/gh-testing-01
    ✓ Deleted Actions secret TEST001 from tinyfists/gh-testing-01
    andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ 
    ```

### `gh variable set`

13. Confirming repo with multiple remotes forces user to choose

   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable set FOO --body "bar"
   multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
   ```
   
14. Confirming setting variable for upstream works as expected

   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable set FOO --body "bar" --repo tinyfists/gh-testing-01 
   ✓ Created variable FOO for tinyfists/gh-testing-01
   ```
   
15. Confirming setting variable for downstream works as expected

   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable set FOO --body "baz" --repo andyfeller/gh-testing-01
   ✓ Created variable FOO for andyfeller/gh-testing-01
   ```

### `gh variable list`
   
16. Confirming repo with multiple remotes forces user to choose

   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable list                                                
   multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
   ```
   
17. Confirming listing variables for upstream works as expected

   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable list --repo tinyfists/gh-testing-01
   NAME  VALUE  UPDATED               
   FOO   bar    less than a minute ago
   ```
   
18. Confirming listing variables for downstream works as expected

   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable list --repo andyfeller/gh-testing-01
   NAME  VALUE  UPDATED               
   FOO   baz    less than a minute ago
   ```

### `gh variable delete`

19. Confirming repo with multiple remotes forces user to choose
   
   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable delete FOO   
   multiple remotes detected [upstream origin]. please specify which repo to use by providing the -R or --repo argument
   ```

17. Confirming deleting variable for downstream works as expected
   
   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable delete FOO --repo tinyfists/gh-testing-01
   ✓ Deleted variable FOO from tinyfists/gh-testing-01
   ```
   
   ```shell 
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable list --repo tinyfists/gh-testing-01     
   no variables found
   ```

18. Confirming upstream variable still exists
   
   ```shell  
   andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ ~/Documents/workspace/cli/cli-trunk/bin/gh variable list --repo andyfeller/gh-testing-01
   NAME  VALUE  UPDATED             
   FOO   baz    about 28 minutes ago
   ```

andyfeller commented:
## Manual testing directly against upstream

Wanting to make sure we were just testing the multiple remotes scenario, this is me running through several of the commands to make sure no regressions.

```shell
andyfeller@Andys-MBP:tinyfists/gh-testing-01 ‹main›$ cd ../../andyfeller
andyfeller@Andys-MBP:workspace/andyfeller $ gh repo clone gh-testing-01
Cloning into 'gh-testing-01'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (3/3), done.
andyfeller@Andys-MBP:workspace/andyfeller $ cd gh-testing-01 
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh variable list
NAME  VALUE  UPDATED         
FOO   baz    about 1 hour ago
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh secret list
NAME     UPDATED         
TEST002  about 1 hour ago
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh secret set TEST003 --body "on the catwalk"
✓ Set Actions secret TEST003 for andyfeller/gh-testing-01
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh secret list                               
NAME     UPDATED               
TEST002  about 1 hour ago
TEST003  less than a minute ago
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh variable set BAR --body "pipe"
✓ Created variable BAR for andyfeller/gh-testing-01
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh variable list                             
NAME  VALUE  UPDATED               
BAR   pipe   less than a minute ago
FOO   baz    about 1 hour ago
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh variable get BAR
pipe
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh variable delete FOO
✓ Deleted variable FOO from andyfeller/gh-testing-01
andyfeller@Andys-MBP:andyfeller/gh-testing-01 ‹main›$ gh secret delete TEST002
✓ Deleted Actions secret TEST002 from andyfeller/gh-testing-01
```

andyfeller reviewed (approved):
Just want to take a moment and say thank you for the monumental lift here, @wingleung ❤️ 

From the automated tests here and the manual testing I commented on in the PR, I think this might be complete, so I'm giving my tentative :shipit: I'd like @williammartin to follow up when he's back on Wednesday next week solely because of how critical getting this right is. 🙇 

Promise its in the home stretch; again thank you for your patience and closing a security concern in the GitHub CLI! ❤️

wingleung commented:
> Just want to take a moment and say thank you for the monumental lift here, @wingleung ❤️

@andyfeller Pleasure was mine, especially because of the insights on high standards and the serene, collaborative environment we maintained throughout, love it 🙏

williammartin reviewed (changes requested):
I think there is some surprising behaviour in this PR as it is. In particular, I would expect the following three scenarios to prompt:

```
gh repo create my-org/my-repo
gh repo fork --org other-org my-org/my-repo
gh repo clone other-org/my-repo
cd my-repo
gh secret set VAL --body KEY
```

```
gh repo create my-org/my-repo
gh repo fork --org other-org my-org/my-repo
gh repo clone other-org/my-repo
cd my-repo
gh secret list
```

```
gh repo create my-org/my-repo
gh repo fork --org other-org my-org/my-repo
gh repo clone other-org/my-repo
cd my-repo
gh secret delete VAL
```

However, they do not right now, and I think this was alluded to here https://github.com/cli/cli/pull/9083#discussion_r1713739684

Erroring and forcing users to provide a `--repo` flag in an interactive environment in some cases but not others seems inconsistent.

Additionally, I'm not convinced we should be modifying the `variable` set of commands. Breaking people's scripts for the sake of security is one thing, breaking them for consistent UX is another. Fortunately this should be easy to back out / and do later if we determine it is worthwhile.

We're going to discuss in our planning meeting later how to get this in. We feel bad that it's been languishing for so long (and I feel bad because Andy has been asking for my review for weeks 😬 ), so just hang in there @wingleung !

### Acceptance Scripts

Based on https://github.com/cli/cli/issues/4688#issuecomment-2355886105 I created a set of acceptance scripts over in https://github.com/cli/cli/commit/0c9b6ed0ad02cc24b7252eccf791ccf7d8cb925b

wingleung commented:
@williammartin thanks for the review!

The purpose was to only show prompts when prompts are desired, in this case we only show it when we're in interactive mode.

Maybe I'm misinterpreting the following documentation

```
# Paste secret value for the current repository in an interactive prompt
$ gh secret set MYSECRET

# Read secret value from an environment variable
$ gh secret set MYSECRET --body "$ENV_VALUE"
```

I was under the impression `gh secret set MYSECRET --body "$ENV_VALUE"` is not an interactive prompt meaning you might use this command in use cases where prompts are not desired and we would rather want to throw an error and enforce a `--repo` flag where it helps make the use case more stable. For example for local scripts, CI, workflows

williammartin commented:
@wingleung, thanks for you patience here, there's been a lot on.

> The purpose was to only show prompts when prompts are desired, in this case we only show it when we're in interactive mode.

That's correct.

> I was under the impression gh secret set MYSECRET --body "$ENV_VALUE" is not an interactive prompt meaning you might use this command in use cases where prompts are not desired and we would rather want to throw an error and enforce a --repo flag where it helps make the use case more stable. For example for local scripts, CI, workflows

But this isn't quite correct. The fundamental misunderstanding is that interactivity is not a property of the flags, but of the environment that is invoking `gh`, and whether there is a TTY attached. So yes, we **require** that in a non-interactive environment, all the necessary flags are provided. If we are in an interactive environment we have the ability to prompt the user. However, even with a TTY attached (and therefore interactivity enabled), I can still provide all the necessary flags to avoid prompting as is the case in:

```
gh secret set MYSECRET --body "$ENV_VALUE"
```

If I run this in my terminal, with more than one remote, `gh` has the capability to prompt, and it should. Otherwise we'd be forcing our interactive users to provide less information just so they can get the prompt.

---

I went round and round a bit on how I wanted to address this and the code changes and in the end I decided to create a [new PR](https://github.com/cli/cli/pull/10209) to replace this one. This is because I wanted the git history in `main` to reflect the changes, but I didn't want the `variable` changes to be included since it would be confusing. I also didn't want to push over your own work. Therefore I liberally reverted, squashed and rebased your commits into a history that should make sense for just your changes. I have ensured that you retain authorship for those commits as seen in https://github.com/cli/cli/pull/10209/commits

---

I also want to make sure to give you some feedback on your code, and why I changed the parts I changed.

In particular, I want to focus on the [following code](https://github.com/cli/cli/pull/9083/files#diff-6d0331daaa499738cf7f050c24667e7236a3c17d336ae9aa3e1979e7c3a12358R205-R219), which was positioned within the `setRun` function.

```go
err = cmdutil.ValidateHasOnlyOneRemote(opts.HasRepoOverride, opts.Remotes)
if err != nil {
    if opts.Interactive {
        selectedRepo, errSelectedRepo := cmdutil.PromptForRepo(baseRepo, opts.Remotes, opts.Prompter)

        if errSelectedRepo != nil {
            return errSelectedRepo
        }

        baseRepo = selectedRepo
    } else {
        return err
    }
}
```

This code is not bad at all; in particular it is clear to read from top to bottom. However, in my experience, this kind of procedural code over time often becomes a maintenance burden because the variation in behaviour becomes a series of conditionals (sometimes nested) that accrete over time.

One of the biggest indicators to me that there might be a better pattern was the injection of `opts.HasRepoOverride` into `cmdutil.ValidateHasOnlyOneRemote`, which is used like [so](https://github.com/cli/cli/pull/9083/files#diff-a2872015d134692ea893f99409c182583209fcb79da72663d2e899ea379184f4R61-R74):

```go
func ValidateHasOnlyOneRemote(hasRepoOverride bool, remotes func() (ghContext.Remotes, error)) error {
	if !hasRepoOverride && remotes != nil {
        ...
	}

	return nil
}
```

I assume this was an attempt to provide a shared location between commands to shortcircuit the validation if the user had provided the `--repo` flag. This is a totally reasonable attempt to avoid duplication in other commands but I suggest that it's better to make this kind of decision early, because tracing the flag through the call stack to determine its impact requires holding context in your head. As a maintainer, I hope to avoid this kind of thing because well, unlike computers, we're pretty bad at remembering all the details.

Instead of having deeply nested conditonals, I look to push the conditionals back up the stack. [Sandi Metz](https://www.youtube.com/watch?v=XXi_FBrZQiU) (I strongly recommend this video) would describe my chosen approach as isolating the behaviour we want to vary, creating _things_ to fulfil those differing behaviours, and injecting the correct smarter _thing_. 

Concretely, the `setRun` function doesn't need any knowledge of how a base repository was chosen, it just needs a way to get the base repo. In this case we have three variations of base repo selection behaviour:
 * Select the repo provided via the `--repo` flag
 * Select the repo from the git remote or error if there are more
 * Select the repo from a prompt if required

With this in mind we are able to push the conditional decision up the [stack](https://github.com/cli/cli/pull/10209/files#diff-6d558c9af27c96323938cbac46519b9c8398b1fe16ab453f4457bca591641529R52-R61):

```go
opts.BaseRepo = f.BaseRepo
if !cmd.Flags().Changed("repo") {
    // If they haven't specified a repo directly, then we will wrap the BaseRepoFunc in one that errors if
    // there might be multiple valid remotes.
    opts.BaseRepo = shared.RequireNoAmbiguityBaseRepoFunc(opts.BaseRepo, f.Remotes)
    // But if we are able to prompt, then we will wrap that up in a BaseRepoFunc that can prompt the user to
    // resolve the ambiguity.
    if opts.IO.CanPrompt() {
        opts.BaseRepo = shared.PromptWhenAmbiguousBaseRepoFunc(opts.BaseRepo, f.IOStreams, f.Prompter)
    }
}
```

We could go further and push this conditional into a single shared location to be used by each command, but I don't believe it provides significantly more value.

Furthermore this provides us with some nice testable [units](https://github.com/cli/cli/pull/10209/files#diff-abb9ac1bfeac1a326ee375b2eff798fe04eb40b16e898b25b1b05f809ec75758R17) that don't require working through the entire command invocation.

---

Again, apologise for such a long wait on this one. I hope you don't mind that I opened a new PR to get this through, I believe that you have been properly credited for your work in the commits.

Thank you for kicking this off and giving us some code to work with, otherwise I think this wouldn't have got any attention for much longer than it already had.

Cheers,

Will

wingleung commented:
@williammartin I want to thank you for the feedback on this PR, it's very clear and gave me a bit more to learn from ([Sandi Metz Talk 👍 ](https://www.youtube.com/watch?v=XXi_FBrZQiU)). Love your refactor of keeping the downstream functions simple and context agnostic by extracting the parameters in the entrypoint of a command ❤️
